src/qservkafka/storage/qserv.py

```diff
@@ -136,14 +137,22 @@ def retry_decorator(
         async def retry_wrapper(
             client: QservClient, *args: P.args, **kwargs: P.kwargs
         ) -> T:
+            retries = 0
             for _ in range(1, max_tries):
                 try:
-                    return await f(client, *args, **kwargs)
+                    result = await f(client, *args, **kwargs)
+                    break
                 except SlackWebException:
                     msg = f"Qserv API call failed, retrying after {delay}s"
                     client.logger.exception(msg)
                     await asyncio.sleep(delay)
-            return await f(client, *args, **kwargs)
+                    retries += 1
+            else:
+                result = await f(client, *args, **kwargs)
+            if retries > 0:
+                event = QservRetryEvent(retries=retries)
+                await client.events.qserv_retry.publish(event)
+            return result
 
         return retry_wrapper
```

Comment:
What about having a `QservApiFailure` event and logging it in the exception handler for every failure? Then I don't think you'd need the for..else to count the failures.  Something like:

```python
async def retry_wrapper(...):
    for _ in range(max_tries):
        try:
            return await(f(client, *args, **kwargs)
        except SlackWebException:
            msg = f"Qserv API call failed, retrying after {delay}s"
            client.logger.exception(msg)
                event = QservApiFailure()
            await client.events.qserv_api_failure.publish(event)
            await asyncio.sleep(delay)

return retry_wrapper
```

Reply:
Oh! Thank you, yes, I got hung up on counting the number of retries but that's silly, I can just send one event per retry. Thank you! I'll make that simplification.